### PR TITLE
Add localization utility functions for use by the core and add-ons

### PR DESCRIPTION
**Pull request by thomasplevy**
## Description

This PR aims to reduce redundant code copied into all of our add-ons which facilitates localization from a "safe" directory where localization files will not be overwritten during plugin updates.

## How has this been tested?

+ New unit tests
+ Upgraded previous localization tests (in the core)
+ Manually tested with physical localization files loaded on a wp install
+ Added localization to the blocks plugin to check methods function properly with js localization (~I need to write unit tests for this method I just realized~)

## Types of changes

+ Developer utilities

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: class-lifterlms.php [hunk 1]
```diff
@@ -388,26 +388,35 @@ public function add_action_links( $links ) {
 	}
 
 	/**
-	 * Load Localization files
+	 * Localize the plugin
 	 *
-	 * The first loaded file takes priority
+	 * Language files files can be found in the following locations (The first loaded file takes priority):
```
Comment (eri-trabiccolo) [inline class-lifterlms.php]:
```suggestion
	 * Language files can be found in the following locations (The first loaded file takes priority):
```

File: includes/functions/llms-functions-l10n.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ */
+function llms_set_script_translations( $handle, $domain, $plugin_dir = null, $language_dir = null ) {
+
+	$plugin_dir   = $plugin_dir ? $plugin_dir : LLMS_PLUGIN_DIR;
+	$language_dir = $language_dir ? $language_dir : 'languages';
+
+	$safe_dir = llms_l10n_get_safe_directory();
+
+	// Check if the file exists in the safe directory (#1) and load it from there if it does, otherwise fall back to the default directories (#2 then #3).
+	$path = load_script_textdomain( $handle, $domain, $safe_dir ) ? $safe_dir : sprintf( '%1$s/%2$s', $plugin_dir, $language_dir );
```
Comment (eri-trabiccolo) [inline includes/functions/llms-functions-l10n.php]: I'm not sure I get this...
who falls back on #2 then #3?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: includes/functions/llms-functions-l10n.php (hunk 2 above)
Comment (thomasplevy) [reply includes/functions/llms-functions-l10n.php]: After writing tests for this method I realized that while it technically works it also technically doesn't. 

See the rewrite (and included tests)

**Comment by thomasplevy** [comment]
@eri-trabiccolo have a look at https://github.com/gocodebox/lifterlms-blocks/pull/70/commits/1b1026d8d0b688506dcb4b83faebc77cbaf11504 for an implementation of the localization in a "real world" scenario

File: tests/phpunit/unit-tests/class-llms-test-assets.php [hunk 3]
```diff
@@ -600,6 +603,57 @@ public function test_register_script_undefined() {
 
 	}
 
+	public function test_register_script_with_translations() {
+
+		// LLMS_PLUGIN_URL gets messed up in the testing environment
```
Comment (eri-trabiccolo) [inline tests/phpunit/unit-tests/class-llms-test-assets.php]:
```suggestion
		// LLMS_PLUGIN_URL gets messed up in the testing environment.
```

File: tests/phpunit/unit-tests/class-llms-test-assets.php [hunk 4]
```diff
@@ -217,7 +217,7 @@ public function test_get() {
 
 	/**
 	 * Test get() method for an undefined asset.
-	 *
+	 *F
```
Comment (eri-trabiccolo) [inline tests/phpunit/unit-tests/class-llms-test-assets.php]:
```suggestion
	 *
```
uck?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
Ok this is quite fine, but doesn't work with the blocks as plugin though.
Activating the blocks plugin, and creating a course results in a blank editor with some warnings logged:

```
[24-Nov-2020 13:32:13 UTC] PHP Warning:  rtrim() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/formatting.php on line 2718
[24-Nov-2020 13:32:13 UTC] PHP Notice:  Array to string conversion in /shared/httpd/mywp/wordpress.git/wp-content/plugins/lifterlms/includes/class-llms-assets.php on line 302
[24-Nov-2020 13:32:13 UTC] PHP Warning:  fopen() expects parameter 1 to be a valid path, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 6109
[24-Nov-2020 13:32:13 UTC] PHP Warning:  strpos() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 6517
[24-Nov-2020 13:32:13 UTC] PHP Warning:  substr() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 6524
[24-Nov-2020 13:32:13 UTC] PHP Warning:  substr() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 2124
[24-Nov-2020 13:32:13 UTC] PHP Notice:  Array to string conversion in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 2128
[24-Nov-2020 13:32:13 UTC] PHP Warning:  dirname() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/plugin.php on line 755
[24-Nov-2020 13:32:13 UTC] PHP Warning:  rtrim() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/formatting.php on line 2718
[24-Nov-2020 13:32:13 UTC] PHP Notice:  Array to string conversion in /shared/httpd/mywp/wordpress.git/wp-content/plugins/lifterlms/includes/class-llms-assets.php on line 302
[24-Nov-2020 13:32:27 UTC] PHP Warning:  rtrim() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/formatting.php on line 2718
[24-Nov-2020 13:32:27 UTC] PHP Notice:  Array to string conversion in /shared/httpd/mywp/wordpress.git/wp-content/plugins/lifterlms/includes/class-llms-assets.php on line 302
[24-Nov-2020 13:32:27 UTC] PHP Warning:  fopen() expects parameter 1 to be a valid path, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 6109
[24-Nov-2020 13:32:27 UTC] PHP Warning:  strpos() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 6517
[24-Nov-2020 13:32:27 UTC] PHP Warning:  substr() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 6524
[24-Nov-2020 13:32:27 UTC] PHP Warning:  substr() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 2124
[24-Nov-2020 13:32:27 UTC] PHP Notice:  Array to string conversion in /shared/httpd/mywp/wordpress.git/wp-includes/functions.php on line 2128
[24-Nov-2020 13:32:27 UTC] PHP Warning:  dirname() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/plugin.php on line 755
[24-Nov-2020 13:32:27 UTC] PHP Warning:  rtrim() expects parameter 1 to be string, array given in /shared/httpd/mywp/wordpress.git/wp-includes/formatting.php on line 2718
[24-Nov-2020 13:32:27 UTC] PHP Notice:  Array to string conversion in /shared/httpd/mywp/wordpress.git/wp-content/plugins/lifterlms/includes/class-llms-assets.php on line 302
```
This is due to the fact that we merge defaults with `array_merge_recursive()` produced something like this:
```
Array
(
    [base] => Array
        (
            [base_file] => Array
                (
                    [0] => /shared/httpd/mywp/wordpress.git/wp-content/plugins/lifterlms-blocks/lifterlms-blocks.php
                    [1] => /shared/httpd/mywp/wordpress.git/wp-content/plugins/lifterlms/lifterlms.php
                )

            [base_url] => Array
                (
                    [0] => http://mywp.loc/wp-content/plugins/lifterlms-blocks/
                    [1] => http://mywp.loc/wp-content/plugins/lifterlms/
                )

            [version] => Array
                (
                    [0] => 1.9.1
                    [1] => 4.8.0
                )

            [suffix] => Array
                (
                    [0] => 
                    [1] => 
                )

            [dependencies] => Array
                (
                )

        )

    [script] => Array
        (
            [translate] => Array
                (
                    [0] => 1
                    [1] => 
                )

            [path] => assets/js
            [extension] => .js
            [in_footer] => 1
        )

    [style] => Array
        (
            [path] => assets/css
            [extension] => .css
            [media] => all
            [rtl] => 1
        )

)
```
So I propose to change this:
https://github.com/gocodebox/lifterlms/blob/ed002a991d56b3f98969a246e0259060f1875773/includes/class-llms-assets.php#L103
into:
```php
		// Merge defaults.
		foreach ( $defaults as $k => $arr ) {
			if ( isset( $this->defaults[ $k ] ) ) {
				$this->defaults[ $k ] = array_merge( $this->defaults[ $k ], $arr );
			}
		}
```
or similar.

**Comment by thomasplevy** [comment]
I was bewildered when I read your comment because I rewrote the method to *properly* merge the way we initially intended it to work and then went to my CLI to see that I had never committed the changes. Sorry!

See above commit (https://github.com/gocodebox/lifterlms/pull/1439/commits/a0cbb9cf1b3d08db53832152d302cb8a9360d026)

**Comment by eri-trabiccolo** [comment]
> I was bewildered when I read your comment because I rewrote the method to _properly_ merge the way we initially intended it to work and then went to my CLI to see that I had never committed the changes. Sorry!

TIL

**Comment by thomasplevy** [comment]
@eri-trabiccolo does this look good to you after that last commit?

**Review by eri-trabiccolo — APPROVED** [review]

**Pull request merged**